Panic: size of error type exceeds size of result shape

### Description

```sh
echo "`wit-bindgen-go version` | `tinygo version` | `wasmtime --version` | `uname -m`"
wit-bindgen-go version v0.6.2 | tinygo version 0.37.0 darwin/arm64 (using go version go1.24.3 and LLVM version 19.1.2) | wasmtime 32.0.0 (d3054950c 2025-04-21) | arm64
```

`wit-bindgen-go` represents the WIT definition `type alias = result<option<tuple<string, string, string>>, tuple<u16, result<u64>, u8>>` with the following type definition: `type Alias cm.Result[OptionTupleStringStringStringShape, cm.Option[[3]string], cm.Tuple3[uint16, cm.Result[uint64, uint64, struct{}], uint8]]`. 

Initializing a data value of type `Alias` causes a panic with the message `result: size of Err type > data type`.

Component model bindings are generated for the following WIT world:
```
world foo {
 ... wasi imports ...

  type alias = result<option<tuple<string, string, string>>, tuple<u16, result<u64>, u8>>;
  export wasi:cli/run@0.2.0;
}
```

The following Go program implements `foo`. It calls the function `cm.Err` to initialize a value of type `Alias`.
```go
func main() {
	var _ foo.Alias = cm.Err[foo.Alias](
		cm.Tuple3[uint16, cm.Result[uint64, uint64, struct{}], uint8]{
			F0: 0,
			F1: cm.OK[cm.Result[uint64, uint64, struct{}]](0),
			F2: 0,
		},
	)
}
```

The program above is compiled to a Wasm component using TinyGo and executed using Wasmtime, 

### Steps to reproduce

Here is a zipped directory that reproduces the bug:
[test-case.zip](https://github.com/user-attachments/files/20270061/test-case.zip)

1. Unzip `test-case.zip`
2. `cd go-err-type-bigger-than-shape`
3. Build and run with `./verify.sh`

### Expected behavior

The program should execute successfully.

### Actual behavior

The program crashes when executed.

I get the following output:
```
panic: result: size of Err type > data type
Error: failed to run main module `foo.wasm`

Caused by:
    0: failed to invoke `run` function
    1: error while executing at wasm backtrace:
           0:   0x5a50 - main!runtime.panicOrGoexit
           1:   0x2e8a - main!runtime._panic
           2:   0xda73 - main!(*go.bytecodealliance.org/cm.result[example.com/reproduce/go-err-type-bigger-than-shape/ns/pkg/foo.OptionTupleStringStringStringShape, go.bytecodealliance.org/cm.Option[[3]string], go.bytecodealliance.org/cm.Tuple3[uint16, go.bytecodealliance.org/cm.Result[uint64, uint64, struct{}], uint8]]).validate[example.com/reproduce/go-err-type-bigger-than-shape/ns/pkg/foo.OptionTupleStringStringStringShape go.bytecodealliance.org/cm.Option[[3]string] go.bytecodealliance.org/cm.Tuple3[uint16, go.bytecodealliance.org/cm.Result[uint64, uint64, struct{}], uint8]]
           3:   0xda11 - main!go.bytecodealliance.org/cm.Err[example.com/reproduce/go-err-type-bigger-than-shape/ns/pkg/foo.Alias example.com/reproduce/go-err-type-bigger-than-shape/ns/pkg/foo.OptionTupleStringStringStringShape go.bytecodealliance.org/cm.Option[[3]string] go.bytecodealliance.org/cm.Tuple3[uint16, go.bytecodealliance.org/cm.Result[uint64, uint64, struct{}], uint8]]
           4:   0x23d1 - main!main.main
           5:   0x2374 - main!internal/wasi/cli/v0.2.0/run.wasmexport_Run
           6:   0x21e8 - main!internal/wasi/cli/v0.2.0/run.wasmexport_Run$gowrapper-wasmexport
           7:   0x17b7 - main!tinygo_launch
           8:   0x22c6 - main!runtime.scheduler
           9:   0x210e - main!internal/wasi/cli/v0.2.0/run.wasmexport_Run#wasmexport
       note: using the `WASMTIME_BACKTRACE_DETAILS=1` environment variable may show more debugging information
    2: wasm trap: wasm `unreachable` instruction executed
```

### Additional context

The Go source program was derived from programs generated by a differential testing framework for WIT binding generators. The WIT definition was derived from a test case produced by wit-smith.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 ### Fixed
 
 - [#306](https://github.com/bytecodealliance/go-modules/issues/306): do not emit incompatible version feature gates when serializing WIT. For example, if a world with version `0.1.0` *includes* a world from another package with a `@since(version = 0.2.0)` feature gate, then strip that feature gate when serializing to WIT if the version is greater than the world’s package version.
+- [#350](https://github.com/bytecodealliance/go-modules/issues/350): correctly align `record` size to the highest alignment of its fields, per [specification](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#element-size).
 
 
 ## [v0.6.2] — 2025-03-16

--- a/cm/result_test.go
+++ b/cm/result_test.go
@@ -387,3 +387,37 @@ func TestIssue344BoolResult(t *testing.T) {
 		t.Errorf("*v.OK(): %v, expected %v", got, want)
 	}
 }
+
+func TestIssue350(t *testing.T) {
+	type Shape struct {
+		_     HostLayout
+		shape [unsafe.Sizeof(Tuple3[uint16, Result[uint64, uint64, struct{}], uint8]{})]byte
+		// Previously, with bug in (*Record).Size() algorithm:
+		// shape [unsafe.Sizeof(Option[[3]string]{})]byte
+	}
+	type O Option[[3]string]
+	type E Tuple3[uint16, Result[uint64, uint64, struct{}], uint8]
+	type T Result[Shape, O, E]
+
+	shapeSize := unsafe.Sizeof(Shape{})
+	errSize := unsafe.Sizeof(E{})
+
+	if errSize > shapeSize {
+		t.Errorf("size of err type (%d) > size of shape type (%d)", errSize, shapeSize)
+
+		var e E
+		base := uintptr(unsafe.Pointer(&e))
+		f0 := uintptr(unsafe.Pointer(&e.F0)) - base
+		f1 := uintptr(unsafe.Pointer(&e.F1)) - base
+		f2 := uintptr(unsafe.Pointer(&e.F2)) - base
+		t.Logf("Offsets: F0: %d F1: %d F2 %d", f0, f1, f2)
+	}
+
+	// _ = Err[T](
+	// 	Err, uint8]{
+	// 		F0: 0,
+	// 		F1: OK[Result[uint64, uint64, struct{}]](0),
+	// 		F2: 0,
+	// 	},
+	// )
+}

--- a/wit/abi.go
+++ b/wit/abi.go
@@ -9,7 +9,7 @@ import (
 // [Canonical ABI] [size], [alignment], and [flat] representation.
 //
 // [Canonical ABI]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md
-// [size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#size
+// [size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#element-size
 // [alignment]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#alignment
 // [flat]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#flattening
 type ABI interface {

--- a/wit/abi_test.go
+++ b/wit/abi_test.go
@@ -99,7 +99,7 @@ func TestDiscriminant(t *testing.T) {
 	}
 }
 
-func TestTypeSize(t *testing.T) {
+func TestTypeSizeAndAlign(t *testing.T) {
 	tests := []struct {
 		name  string
 		v     Type
@@ -119,6 +119,10 @@ func TestTypeSize(t *testing.T) {
 		{"f64", F64{}, 8, 8},
 		{"char", Char{}, 4, 4},
 		{"string", String{}, 8, 4},
+		{"option<string>", &TypeDef{Kind: &Option{Type: String{}}}, 12, 4},
+		{"option<f32>", &TypeDef{Kind: &Option{Type: F32{}}}, 8, 4},
+		{"variant", &TypeDef{Kind: &Variant{Cases: []Case{{Type: String{}}, {Type: F64{}}}}}, 16, 8},
+		{"record", &TypeDef{Kind: &Record{Fields: []Field{{Type: U16{}}, {Type: &TypeDef{Kind: &Result{OK: U64{}}}}, {Type: U8{}}}}}, 32, 8},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
@@ -156,6 +160,7 @@ func TestTypeFlat(t *testing.T) {
 		{"option<string>", &TypeDef{Kind: &Option{Type: String{}}}, []Type{U32{}, PointerTo(U8{}), U32{}}},
 		{"option<f32>", &TypeDef{Kind: &Option{Type: F32{}}}, []Type{U32{}, F32{}}},
 		{"variant", &TypeDef{Kind: &Variant{Cases: []Case{{Type: String{}}, {Type: F64{}}}}}, []Type{U32{}, U64{}, U32{}}},
+		{"record", &TypeDef{Kind: &Record{Fields: []Field{{Type: U16{}}, {Type: &TypeDef{Kind: &Result{OK: U64{}}}}, {Type: U8{}}}}}, []Type{U32{}, U32{}, U64{}, U32{}}},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
@@ -177,9 +182,6 @@ func witFor[T Node](nodes ...T) []string {
 
 // TestHasBorrow verifies that HasBorrow returns true for WIT types that contain a Borrow type.
 func TestHasBorrow(t *testing.T) {
-	makeBorrow := func() *TypeDef { return &TypeDef{Kind: &Borrow{}} }
-	makeTypeDef := func(kind TypeDefKind) *TypeDef { return &TypeDef{Kind: kind} }
-
 	testCases := []struct {
 		name     string
 		typeDef  *TypeDef
@@ -210,4 +212,12 @@ func TestHasBorrow(t *testing.T) {
 			}
 		})
 	}
+}
+
+func makeTypeDef(kind TypeDefKind) *TypeDef {
+	return &TypeDef{Kind: kind}
+}
+
+func makeBorrow() *TypeDef {
+	return &TypeDef{Kind: &Borrow{}}
 }

--- a/wit/enum.go
+++ b/wit/enum.go
@@ -29,7 +29,7 @@ func (e *Enum) Despecialize() TypeDefKind {
 // type that can represent 0...len(e.Cases).
 // It is first [despecialized] into a [Variant] with no associated types, then sized.
 //
-// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#size
+// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#element-size
 // [despecialized]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#despecialization
 func (e *Enum) Size() uintptr {
 	return e.Despecialize().Size()

--- a/wit/flags.go
+++ b/wit/flags.go
@@ -11,7 +11,7 @@ type Flags struct {
 
 // Size returns the [ABI byte size] of [Flags] f.
 //
-// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#size
+// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#element-size
 func (f *Flags) Size() uintptr {
 	n := len(f.Flags)
 	switch {

--- a/wit/future.go
+++ b/wit/future.go
@@ -12,7 +12,7 @@ type Future struct {
 
 // Size returns the [ABI byte size] for a [Future].
 //
-// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#size
+// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#element-size
 func (*Future) Size() uintptr { return 4 }
 
 // Align returns the [ABI byte alignment] a [Future].

--- a/wit/list.go
+++ b/wit/list.go
@@ -11,7 +11,7 @@ type List struct {
 
 // Size returns the [ABI byte size] for a [List].
 //
-// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#size
+// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#element-size
 func (*List) Size() uintptr { return 8 } // [2]int32
 
 // Align returns the [ABI byte alignment] a [List].

--- a/wit/option.go
+++ b/wit/option.go
@@ -27,7 +27,7 @@ func (o *Option) Despecialize() TypeDefKind {
 // Size returns the [ABI byte size] for [Option] o.
 // It is first [despecialized] into a [Variant] with two cases, "none" and "some(T)", then sized.
 //
-// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#size
+// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#element-size
 // [despecialized]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#despecialization
 func (o *Option) Size() uintptr {
 	return o.Despecialize().Size()

--- a/wit/pointer.go
+++ b/wit/pointer.go
@@ -14,7 +14,7 @@ type Pointer struct {
 
 // Size returns the [ABI byte size] for [Pointer].
 //
-// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#size
+// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#element-size
 func (*Pointer) Size() uintptr { return 4 }
 
 // Align returns the [ABI byte alignment] for [Pointer].

--- a/wit/record.go
+++ b/wit/record.go
@@ -11,7 +11,7 @@ type Record struct {
 
 // Size returns the [ABI byte size] for [Record] r.
 //
-// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#size
+// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#element-size
 func (r *Record) Size() uintptr {
 	var s uintptr
 	for _, f := range r.Fields {

--- a/wit/record.go
+++ b/wit/record.go
@@ -18,7 +18,7 @@ func (r *Record) Size() uintptr {
 		s = Align(s, f.Type.Align())
 		s += f.Type.Size()
 	}
-	return s
+	return Align(s, r.Align())
 }
 
 // Align returns the [ABI byte alignment] for [Record] r.

--- a/wit/resource.go
+++ b/wit/resource.go
@@ -8,7 +8,7 @@ type Resource struct{ _typeDefKind }
 
 // Size returns the [ABI byte size] for [Resource].
 //
-// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#size
+// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#element-size
 func (*Resource) Size() uintptr { return 4 }
 
 // Align returns the [ABI byte alignment] for [Resource].
@@ -45,7 +45,7 @@ func (_handle) isHandle() {}
 
 // Size returns the [ABI byte size] for this [Handle].
 //
-// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#size
+// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#element-size
 func (_handle) Size() uintptr { return 4 }
 
 // Align returns the [ABI byte alignment] for this [Handle].

--- a/wit/result.go
+++ b/wit/result.go
@@ -40,7 +40,7 @@ func (r *Result) Types() []Type {
 // Size returns the [ABI byte size] for [Result] r.
 // It is first [despecialized] into a [Variant] with two cases "ok" and "error", then sized.
 //
-// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#size
+// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#element-size
 // [despecialized]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#despecialization
 func (r *Result) Size() uintptr {
 	return r.Despecialize().Size()

--- a/wit/stream.go
+++ b/wit/stream.go
@@ -12,7 +12,7 @@ type Stream struct {
 
 // Size returns the [ABI byte size] for a [Stream].
 //
-// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#size
+// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#element-size
 func (*Stream) Size() uintptr { return 4 }
 
 // Align returns the [ABI byte alignment] a [Stream].

--- a/wit/tuple.go
+++ b/wit/tuple.go
@@ -48,7 +48,7 @@ func (t *Tuple) Despecialize() TypeDefKind {
 // Size returns the [ABI byte size] for [Tuple] t.
 // It is first [despecialized] into a [Record] with 0-based integer field names, then sized.
 //
-// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#size
+// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#element-size
 // [despecialized]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#despecialization
 func (t *Tuple) Size() uintptr {
 	return t.Despecialize().Size()

--- a/wit/variant.go
+++ b/wit/variant.go
@@ -46,7 +46,7 @@ func (v *Variant) Types() []Type {
 
 // Size returns the [ABI byte size] for [Variant] v.
 //
-// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#size
+// [ABI byte size]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#element-size
 func (v *Variant) Size() uintptr {
 	s := Discriminant(len(v.Cases)).Size()
 	s = Align(s, v.maxCaseAlign())

